File: hmailserver/source/Server/Common/Persistence/Maintenance/Maintenance.cpp
// Copyright (c) 2009 Martin Knafve / hMailServer.com.  
// https://www.progressiverobot.com
// Copyright (c) 2026 Christopher Holloway / Progressive Robot Ltd

#include "stdafx.h"
#include "Maintenance.h"

#ifdef _DEBUG
#define DEBUG_NEW new(_NORMAL_BLOCK, __FILE__, __LINE__)
#define new DEBUG_NEW
#endif

namespace HM   
{
   Maintenance::Maintenance()
   {

   }

   Maintenance::~Maintenance()
   {

   }

   bool
   Maintenance::Perform(MaintenanceOperation operation)
   {
      switch (operation)
      {
      case RecalculateFolderUID:
         return RecalculateFolderUID_();
      }

      return false;
   }

   // Goes through all mailboxes and sets the foldercurrentuid to the latest message uid.
   bool
   Maintenance::RecalculateFolderUID_()
   {
      /*
         The rows this repair is about are messages that live in an IMAP folder.
         hm_messages holds more than that: a message awaiting delivery is stored
         with messagefolderid = 0 and messageuid = 0 (Message.cpp's defaults),
         because it is in the queue rather than in anyone's mailbox.

         The query used to group the whole table and then bail out with `return
         false` on the first row whose folder or uid was not positive - so on any
         server with even one message in the delivery queue, the grouped set
         contained a (0, 0) row and the repair failed. Worse than failing: the
         result set has no ORDER BY, so an arbitrary number of folders had already
         been updated before the abort. The run was neither complete nor a
         no-op, and the caller was told only "The maintenance operation failed".

         Excluding those rows in SQL is the fix. It is also more correct than the
         guards were: they treated ordinary queue contents as corruption, when the
         only genuinely bad state - a folder id or uid that is negative - is
         excluded by the same predicate.
      */
      AnsiString recordSQL =
         "SELECT messagefolderid, MAX(messageuid) as messageuid FROM hm_messages "
         "WHERE messagefolderid > 0 AND messageuid > 0 GROUP BY messagefolderid";

      std::shared_ptr<DALRecordset> pRS = Application::Instance()->GetDBManager()->OpenRecordset(SQLCommand(recordSQL));

      if (!pRS)
         return false;

      while (!pRS->IsEOF())
      {
         __int64 messageFolderID = pRS->GetInt64Value("messagefolderid");
         __int64 messageUID = pRS->GetInt64Value("messageuid");

         // Belt and braces: the predicate above already excludes these, so a row
         // reaching here means the database returned something the query said it
         // would not. Skip it rather than abandoning the run, because abandoning
         // half way is what made the old behaviour worse than useless.
         if (messageFolderID <= 0 || messageUID <= 0)
         {
            pRS->MoveNext();
            continue;
         }

         AnsiString sqlUpdate = Formatter::Format("UPDATE hm_imapfolders SET foldercurrentuid = {0} WHERE folderid = {1} AND foldercurrentuid < {0}", messageUID, messageFolderID);

         bool result = Application::Instance()->GetDBManager()->Execute(SQLCommand(sqlUpdate));
         if (result == false)
            return false;

         pRS->MoveNext();
      }

      return true;
   }

}

File: hmailserver/test/RegressionTests/Infrastructure/Persistence/Maintenance.cs
// Copyright (c) 2010 Martin Knafve / hMailServer.com.  
// https://www.progressiverobot.com
// Copyright (c) 2026 Christopher Holloway / Progressive Robot Ltd

using System.Runtime.InteropServices;
using hMailServer;
using NUnit.Framework;
using NUnit.Framework.Legacy;
using RegressionTests.Shared;

namespace RegressionTests.Infrastructure.Persistence
{
   [TestFixture]
   public class Maintenance : TestFixtureBase
   {
      [Test]
      public void TestUnknownOperation()
      {
         var ex = Assert.Throws<COMException>(() =>
            _application.Utilities.PerformMaintenance((eMaintenanceOperation)234));
         StringAssert.Contains("Unknown maintenance operation.", ex.Message);
      }

      [Test]
      public void TestUpdateIMAPFolderUID()
      {
         // Set up a basic environment which we can work with.
         var backupRestore = new BackupRestore();
         backupRestore.SetUp();
         backupRestore.SetupEnvironment();

         _application.Utilities.PerformMaintenance(eMaintenanceOperation.eUpdateIMAPFolderUID);
      }

      /// <summary>
      /// The repair must survive mail sitting in the delivery queue.
      ///
      /// RecalculateFolderUID_ grouped every row of hm_messages and returned false
      /// on the first row whose folder or uid was not positive. A message awaiting
      /// delivery is stored with messagefolderid = 0 and messageuid = 0, so on any
      /// server with a queue the grouped set contained a (0, 0) row and the whole
      /// operation failed - after updating an arbitrary number of folders, because
      /// the query has no ORDER BY. Neither complete nor a no-op, and reported only
      /// as "The maintenance operation failed".
      ///
      /// The queued row is produced by routing a domain at a closed port on
      /// 127.0.0.1: the connection is refused immediately, so the message stays in
      /// hm_messages with folder id 0 and uid 0, with no DNS lookup and no
      /// dependence on the network. Against the previous build the maintenance call
      /// throws a COMException.
      /// </summary>
      [Test]
      public void TestUpdateIMAPFolderUIDWithMessagesInTheDeliveryQueue()
      {
         var account = SingletonProvider<TestSetup>.Instance.AddAccount(_domain, "uidmaint@example.test", "test");

         // One message that IS in a folder, so the repair has something to do and
         // the assertion at the end means something.
         var smtpClient = new SmtpClientSimulator();
         smtpClient.Send(account.Address, account.Address, "In a folder", "Body");
         ImapClientSimulator.AssertMessageCount(account.Address, "test", "Inbox", 1);

         try
         {
            // And one that is not. A route sends this domain to a port nothing is
            // listening on, so the delivery attempt is refused at once and the
            // message stays in the queue - a row with folder id 0 and uid 0, which
            // is exactly what used to abort the whole run.
            var route = _settings.Routes.Add();
            route.DomainName = "queued.example.test";
            route.TargetSMTPHost = "127.0.0.1";
            route.TargetSMTPPort = 1;
            // Enough tries that the failure does not retire the message, and a
            // retry interval long enough that it will not be attempted again
            // while this test runs - it has to still be in the queue when the
            // maintenance operation is called.
            route.NumberOfTries = 5;
            route.MinutesBetweenTry = 5;
            route.Save();

            smtpClient.Send(account.Address, "someone@queued.example.test", "Queued", "Body");

            CustomAsserts.AssertRecipientsInDeliveryQueue(1, false);

            // The whole point: this must not throw.
            Assert.DoesNotThrow(() =>
               _application.Utilities.PerformMaintenance(eMaintenanceOperation.eUpdateIMAPFolderUID));

            // And the folder that did have a message must have been brought up to
            // date rather than skipped - the old code could abort before reaching it.
            var folder = account.IMAPFolders.get_ItemByName("Inbox");
            ClassicAssert.GreaterOrEqual(folder.CurrentUID, 1);
         }
         finally
         {
            // PerformBasicSetup does not clear the delivery queue, so a message left
            // here would be inherited by whichever fixture runs next. Routes it does
            // remove, but the queue is this test's to clean up - and so is anything
            // the refused connection put in the error log, or the next fixture's
            // AssertNoReportedError fails on this test's deliberate failure.
            TestSetup.DeleteMessagesInQueue();
            LogHandler.DeleteCurrentDefaultLog();
            LogHandler.ClearErrorLogUntilSettled();
         }
      }
   }
}